pandas/tseries/offsets.py

```diff
@@ -1160,6 +1161,104 @@ def onOffset(self, dt):
     _prefix = 'MS'
 
 
+class SemiMonthOffset(SingleConstructorOffset):
+    _adjust_dst = True
+
+    @apply_wraps
+    def apply(self, other):
+        n = self.n
+        if n >= 0:
+            if n == 0 and not self.onOffset(other):
+                n = 1  # rollforward
+            for _ in range(n):
+                _, days_in_month = tslib.monthrange(other.year, other.month)
+                if 1 <= other.day < 15:
+                    other += relativedelta(day=15)
+                elif 15 <= other.day < days_in_month:
+                    other = self._rollforward_from_15(other)
+                else:
+                    other = self._rollforward_from_31(other)
+        else:
+            for _ in range(abs(n)):
+                _, days_in_month = tslib.monthrange(other.year, other.month)
+                if 15 < other.day <= days_in_month:
+                    other += relativedelta(day=15)
+                elif 1 < other.day <= 15:
+                    other = self._rollback_from_15(other)
+                else:
+                    other = self._rollback_from_1(other)
+        return other
+
+    def _rollforward_from_15(self, other):
+        """Return the next offset if other is on the 15th of the month"""
+        raise NotImplementedError
+
+    def _rollback_from_15(self, other):
+        """Return the previous offset if other is on the 15th of the month"""
+        raise NotImplementedError
+
+    def _rollback_from_1(self, other):
+        """Return the previous offset if other is on the 1st of the month"""
+        raise NotImplementedError
+
+    def _rollforward_from_31(self, other):
+        """Return the next offset if other is on the last day of the month"""
+        raise NotImplementedError
+
+
+class SemiMonthEnd(SemiMonthOffset):
+    """
+    DateOffset of a half month on the 15h and end of the month
+
+    .. versionadded:: 0.18.2
+    """
+    _prefix = 'SM'
+
+    def onOffset(self, dt):
+        if self.normalize and not _is_normalized(dt):
+            return False
+        _, days_in_month = tslib.monthrange(dt.year, dt.month)
+        return dt.day in (15, days_in_month)
+
+    def _rollback_from_15(self, other):
```

Comment:
you can move all of the rollforward/back to the parent class to avoid duplication I believe